src/vehicle.cpp

```diff
@@ -4980,105 +4981,155 @@ std::string aim_type( const vehicle_part &part )
-        if( fuel_left( part_info( p ).fuel_type ) < 1 && ( items.empty() || items.front().charges < 1 ) ) {
-            aimed = _("No ammo");
-            en = false;
-        } else {
-            aimed = aim_type( parts[p] );
-            en = true;
-        }
-        
-        pmenu.addentry( i++, en, MENU_AUTOASSIGN, _("[%s] %s"),
-                        aimed.c_str(), part_info( p ).name.c_str() );
+    const item gun( part_info( p ).item, 0 );
+    itype *am_itype = nullptr;
+    auto items = get_items( p );
+    if( !items.empty() && items.front().charges > 0 ) {
+        am_itype = items.front().type;
+    } else if( gun.is_charger_gun() ) { // Charger guns "use" different ammo than they fire
+        am_itype = item::find_type( "charge_shot" );
+    } else if( fuel_left( part_info( p ).fuel_type ) > 0 ) {
+        am_itype = item::find_type( part_info( p ).fuel_type );
+    } else if( gun.has_flag( "NO_AMMO" ) ) {
+        am_itype = item::find_type( "generic_no_ammo" );
+    } else {
+        // Couldn't find ammo
+        return -1;
     }
 
-    pmenu.addentry( i, true, 'q', _("Finish") );
-    pmenu.w_y = 0; // Move the menu so that we can see our vehicle
+    const auto ammo = am_itype->ammo.get();
+    const auto &gun_data = *gun.type->gun;
+    return gun_data.range + ( ammo != nullptr ? ammo->range : 0 );
+}
+
+turret_fire_ability vehicle::turret_can_shoot( const int p, const point &pos )
+{
+    if( part_flag( p, "MANUAL" ) ) {
+        return turret_wont_aim;
+    }
+
+    if( parts[p].mode == 0 ) {
+        return turret_is_off;
+    }
+
+    if( turret_has_ammo( p ) < 1 ) {
+        return turret_no_ammo;
+    }
+
+    const int turrange = get_turret_range( p, true );
+    point tpos( global_pos() + parts[p].precalc[0] );
+    if( rl_dist( tpos, pos ) > turrange ) {
+        return turret_out_of_range;
+    }
     
-    pmenu.selected = selected;
-    pmenu.fselected = selected;
-    pmenu.query();
-    if( pmenu.ret < 0 || pmenu.ret >= i ) {
-        return;
+    return turret_all_ok;
+}
+
+std::map< int, turret_fire_ability > vehicle::turrets_can_shoot( const point &pos )
+{
+    std::vector< int > turrets = all_parts_with_feature( "TURRET", true );
+    std::map< int, turret_fire_ability > ret;
+    for( const int turret_index : turrets ) {
+        ret[turret_index] = turret_can_shoot( turret_index, pos );
     }
 
-    selected = pmenu.ret;
+    return ret;
+}
 
-    int turret_index = turrets[selected];
+bool vehicle::aim_turrets()
+{
+    std::vector< int > turrets = all_parts_with_feature( "TURRET", true );
+    // We only want turrets that can aim by themselves and are on
+    auto cant_shoot = [&]( const int tur ) {
+        return part_flag( tur, "MANUAL" ) || parts[tur].mode == 0;
+    };
+    turrets.erase( std::remove_if( turrets.begin(), turrets.end(), cant_shoot ), turrets.end() );
+    if( turrets.empty() ) {
+        add_msg( m_warning, _("Can't aim turrets: all turrets are offline") );
```

Comment:
I'm probably just doing something wrong, but I can't trigger anything but this message, e.g. trying to invoke it from several different APCs.  Also I can't toggle the modes of the individual turrets, they're stuck in off I think, even though they say auto.
Does this only interact with a particular type of turret?